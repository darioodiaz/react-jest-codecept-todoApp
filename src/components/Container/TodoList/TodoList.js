import React, { Component, PropTypes } from 'react';
import TodoItem from './TodoItem/TodoItem.js';
import NewTodoItem from './NewTodoItem/NewTodoItem.js';

class TodoList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            todos: props.todos,
            newTodo: ''
        };
        this.onNewTodo = this.onNewTodo.bind(this);
        this.onSelectItem = this.onSelectItem.bind(this);
    }
    onSelectItem(id) {
        this.props.onSelectItem(id);
    }
    getNewTodo() {
        return {
            text: this.state.newTodo,
            selected: false
        };
    }
    clearNewTodo() {
        this.refs.newTodo.clear();
        this.setState({ newTodo: '' });
    }
    onNewTodo(todo) {
        this.setState({ newTodo: todo });
    }
    componentWillReceiveProps(nextProps) {
        this.setState({ todos: nextProps.todos });
    }

    render() {
        return (
            <ul className="todo-list">
                {this.state.todos.map((todo, i) => <TodoItem onSelectItem={this.onSelectItem} id={todo.id} key={todo.id} todo={todo.text} selected={todo.selected} />)}
                <NewTodoItem ref="newTodo" onNewTodo={this.onNewTodo} />
            </ul>
        );
    }
}

TodoList.propTypes = {
    todos: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        text: PropTypes.string.isRequired,
        selected: PropTypes.bool.isRequired
    })).isRequired,
    onSelectItem: PropTypes.func.isRequired,
    newTodo: PropTypes.string
};

export default TodoList;