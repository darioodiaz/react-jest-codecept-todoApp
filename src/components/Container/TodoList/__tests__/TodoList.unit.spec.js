import React from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils from 'react-addons-test-utils';
import TodoList from '../TodoList';

it('montar el componente sin problemas', () => {
    const todos = [{ id: 0, text: 'Todo 0', selected: false }];
    shallow(<TodoList todos={todos} onSelectItem={jest.fn()} />);
});

it('tener la misma cantidad de todos', () => {
    const todos = [{ id: 0, text: 'Todo 0', selected: false }, { id: 1, text: 'Todo 1', selected: false }];
    const el = shallow(<TodoList todos={todos} onSelectItem={jest.fn()} />);
    const _todos = el.state().todos;
    expect(toJson(el)).toMatchSnapshot();
    _todos.forEach((item, i) => {
        expect(item).toEqual(todos[i]);
    });
});

it('llamar a la funcion onSelectItem', () => {
    const todos = [{ id: 0, text: 'Todo 0', selected: false }];
    const onSelectItem = jest.fn();
    const el = shallow(<TodoList todos={todos} onSelectItem={onSelectItem} />);
    el.instance().onSelectItem(0);
    expect(onSelectItem).toHaveBeenCalledWith(0);
});

it('cambiar los todos y tener los mismos cambiados', () => {
    const todos = [{ id: 0, text: 'Todo 0', selected: false }];
    const newTodos = [{ id: 0, text: 'Todo 0', selected: false }, { id: 1, text: 'Todo 1', selected: false }];
    const el = mount(<TodoList todos={todos} onSelectItem={jest.fn()} />);
    expect(el.state().todos).toEqual(todos);
    el.setProps({ todos: newTodos });
    expect(el.state().todos).toEqual(newTodos);
});

it('tener un nuevo todo y llamar a la funcion clearTodo y borrar el estado', () => {
    const todos = [{ id: 0, text: 'Todo 0', selected: false }];
    const onSelectItem = jest.fn();
    const el = mount(<TodoList todos={todos} onSelectItem={onSelectItem} />);
    el.instance().onNewTodo('Todo1');
    expect(el.state().newTodo).toBe('Todo1');
    el.instance().clearNewTodo();
    expect(el.state().newTodo).toBe('');
});

it('tener un nuevo todo con texto Todo', () => {
    const todos = [{ id: 0, text: 'Todo 0', selected: false }, { id: 1, text: 'Todo 1', selected: false }];
    const el = shallow(<TodoList todos={todos} onSelectItem={jest.fn()} onNewTodo={jest.fn()} />);
    el.instance().onNewTodo('Todo');
    const todo = el.state().newTodo;
    expect(todo).toBe('Todo');
});

it('tener un nuevo todo con texto Todo y no estar seleccionado', () => {
    const todos = [{ id: 0, text: 'Todo 0', selected: false }, { id: 1, text: 'Todo 1', selected: false }];
    const el = shallow(<TodoList todos={todos} onSelectItem={jest.fn()} onNewTodo={jest.fn()} />);
    const todo = { text: 'Todo', selected: false };
    el.instance().onNewTodo(todo.text);
    const _todo = el.instance().getNewTodo();
    expect(_todo).toEqual(todo);
});