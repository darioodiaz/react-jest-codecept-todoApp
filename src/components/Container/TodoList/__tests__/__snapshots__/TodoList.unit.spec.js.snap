exports[`test tener la misma cantidad de todos 1`] = `
<ul
  className="todo-list">
  <TodoItem
    id={0}
    onSelectItem={[Function]}
    selected={false}
    todo="Todo 0" />
  <TodoItem
    id={1}
    onSelectItem={[Function]}
    selected={false}
    todo="Todo 1" />
  <NewTodoItem
    onNewTodo={[Function]} />
</ul>
`;
